### PR TITLE
Merkle

### PR DESCRIPTION
**Pull request by barnjamin**

### CODE REVIEW

File: python/fp.py [hunk 1]
```diff
@@ -60,8 +66,25 @@ def __div__(self, other: "ExtElem") -> "ExtElem":
...
+                n = other.n
+            case _:
+                raise Exception("??")
+
+        tot = ExtElemOne
+        x = self
+        while n != 0:
+            if n % 2 == 1:
+                tot *= x
+            n = int(n / 2)
```
Comment (tzaffi) [inline python/fp.py:85]:
```suggestion
            n = n // 2
```

File: python/fp.py [hunk 2]
```diff
@@ -36,20 +46,65 @@ def __init__(self, e: list[Elem]):
...
 
+    def inv(self) -> "ExtElem":
+        a = self.e
+        # Compute the multiplicative inverse by looking at `ExtElem` as a composite
+        # field and using the same basic methods used to invert complex
+        # numbers. We imagine that initially we have a numerator of `1`, and a
+        # denominator of `a`. `out = 1 / a`; We set `a'` to be a with the first
+        # and third components negated. We then multiply the numerator and the
+        # denominator by `a'`, producing `out = a' / (a * a')`. By construction
+        # `(a * a')` has `0`s in its first and third elements. We call this
```
Comment (ahangsu) [inline python/fp.py:57]: should be 0's in second and fourth element? i.e., x and x^3 term.

Comment (ahangsu) [reply python/fp.py:62]: nvm, I think these folks start from zeroist.

File: python/fp.py [hunk 3]
```diff
@@ -17,15 +20,27 @@ def __add__(self, other: "Elem") -> "Elem":
...
-    def __pow__(self, other: "Elem") -> "Elem":
-        return Elem(pow(self.n, other.n))
+    def __pow__(self, other: "Elem | int") -> "Elem":
+        match other:
+            case int():
+                return Elem(pow(self.n, other))
+            case Elem():
+                return Elem(pow(self.n, other.n))
+            case _:
+                raise Exception("??")
```
Comment (ahangsu) [inline python/fp.py:30]: I doubt these lines are correct, for they are montgomery encoded, so `self.n` has $R$ attached to itself.

**Pull request merged**